variant/variant.go

```diff
@@ -0,0 +1,55 @@
+package variant
+
+import (
+	"github.com/vertgenlab/gonomics/dna"
+)
+
+// Substitution describes a single base change.
+// Pos is the 0-base position of the changed base.
+// (e.g. ATG sub 1 T>C -> ACG)
+type Substitution struct {
+	Chr string
+	Pos int
+	Ref dna.Base
+	Alt dna.Base
+}
+
+// Insertion describes an insertion into a sequence.
+// Pos is the 0-base position of the base after the insertion.
+// (e.g. ATG ins 1 C -> ACTG)
+type Insertion struct {
+	Chr string
+	Pos int
+	Seq []dna.Base
+}
+
+// Deletion describes a sequence deletion.
+// Start and End are the 0-base left-closed right-open interval
+// being deleted (e.g. ATG del [1,2) -> AG).
+type Deletion struct {
+	Chr   string
+	Start int
+	End   int
+}
+
+// Delins describes a combined insertion and a deletion of
+// a sequence. Useful for describing complex variants.
+// Start and End are the 0-base left-closed right-open interval
+// being deleted (e.g. ATG del [1,3) ins C -> AC).
+type Delins struct {
```

Comment:
I am not against the Delins, but why not leave it as a deletion followed by an insertion, or an insertion followed by a deletion?  Off the top of my head, I can't think of a special case why we would want to pair these two events and treat them as one, rather than keeping them as separate events where one happens after the other.  I am thinking about a slice of variants where the order of the variants in a slice would describe the order that those variants should be applied to the starting sequence.  We would again run into the question of if variants after the initial insertion/deletion/delins would use the "original" or "current" coordinate system.  Maybe the "current" system so that we could mutate a base that had been inserted and was not in the original sequence?

Reply:
When i initially wrote this I did in fact have Delins as just a Deletion followed by an Insertion. This worked fine for Mutate, but it complicates Effect quite a bit. As a result I decided to split it into its own type. This not only makes Effect prediction a bit easier, it also allows for several small efficiencies that are not possible when doing a deletion and insertion in series. It was also partly motivated by HGVS guidelines (https://varnomen.hgvs.org/recommendations/DNA/variant/delins/). They were not necessarily designed with programing in mind, but they recognize it as a distinct class of mutations.